Made quizzes async and used new delighters (#1)
codecademy-engineering/frontend-takehome: README.md

## The Situation

Codecademy needs an embeddable multiple choice quiz widget to help learners evaluate their skills.

> It should be viewable on both wide ("desktop") and narrow ("mobile") screens.

Reviewer: If we remove mobile screens, this will need to be updated

We do mean _two_: completing additional features will _not_ gain you points!
Please only complete _two_.

> - A: Once both of the quizzes have gotten all answers correct in a session at least once each, use `getMoreQuizzes` to load in more quizzes that the user can take.

Reviewer: ```suggestion
- A: Once both of the quizzes have gotten all answers correct, use `getMoreQuizzes` to load in more quizzes that the user can take.
```

Suggestion to make it more concise:
"in a session" seems unnecessary
"at least once" seems redundant since they can't answer it more than once anyway

> - B: In addition to the "Next" button after taking a quiz, add a "Retake" button and a summary giving how many times the quiz has been taken.

Author: ```suggestion
- B: In addition to the "Next" button after taking a quiz, add a "Retake" button and a display the number of times the quiz has been taken.
```

> - B: In addition to the "Next" button after taking a quiz, add a "Retake" button and a number indicating the times it has ever been completed.

Reviewer: So given the following behavior:
```
Take quiz 1
retake quiz 1
take quiz 2
take quiz 1
```

After completing quiz 1 the last time, the expectation is it would show 3 retakes right?

I feel like maybe this may be clearer for that?

```suggestion
- B: In addition to the "Next" button after taking a quiz, add a "Retake" button and a summary giving how many times the quiz has been taken.
```

> - C: In addition to the "Next" button after taking a quiz, display a list of the quiz's questions with the user's selected answer beside each question.

Reviewer: ```suggestion
- C: In addition to the summary showing the number of questions correct after taking a quiz, display a list of the quiz's questions with the user's selected answer beside each question.
```

This delighter is adding onto the question summary, not the next button

> - D: [Add tests](https://codesandbox.io/docs/tests).

Reviewer 2: Judah had a suggestion for this, about potentially adding "use Gamut" as a delighter; might be a better future iteration.

> See `./mockups/delighters/*.png` for rough descriptions of what A, B, and C would look like.

Reviewer 3: do we have these mockups? i'm still kind of confused what B and C look like

Author: @borisonr I was planning on waiting until we're ok with them before going through the effort 😄

Reviewer 3: totally fair, maybe if everyone is good with these, after i see the mockups i can give the wording a shot? i'm mostly confused about C. Do you mean it will look something like:

Good job!

This is what you answered:
1. image q
A: <img/>
2. css q
A: .pretty

{Next}

Author: Exactly that, yes!

Reviewer 3: small tweak but how about `In addition to the "Next" button after taking a quiz, display a list of the quiz's questions with the user's selected answer beside each question.`

Author: Applied 🌟

We do mean _two_: completing additional features will _not_ gain you points!
Please only complete _two_.

> - A: Display a progress bar on the bottom of the component. It should fill in with green from the left for the percentage of questions correctly answered, then red for the percentage of questions answered incorrectly, then white for the remaining not-yet-answered questions.

Reviewer: I like removing this one - all the answers from candidates have been the same. I haven't seen someone do bad poorly on this yet

> - A: When a quiz is completed, show a summary of its questions and choices, along with how long was spent on each question before answering.

Reviewer: I like the summary page showing quiz answers - that tests handling state and passing it down.

I'm not sure if we need the time one - what skills is that meant to test?

Author: Yeah I'll just remove it.

> - B: In addition to the "Next" button after taking a quiz, add a "Retake" button. After retaking a quiz, its summary screen should also display the previous attempt's information.

Reviewer 3: if you do B you'd have to also do A...

> - B: In addition to the "Next" button after taking a quiz, add a "Retake" button. The quiz summary screen should display all previous scores for that quiz.

Reviewer: I feel like this one won't play well with A. It'll make for a complex summary page if they decide to do both

Author: Hmmm... changed it to just the one previous attempt's information. Thoughts?

> - C: Once both of the quizzes have gotten all answers correct in a session at least once each, use `getMoreQuizzes` to load in more quizzes that the user can take.

Reviewer 3: you'd have to do B to do this one because you need to be able to retake them, which means you'd also have to do A, which is 3 delighters..?

Author: _It's a trap!_

Just kidding - these are poorly/incorrectly worded. I'll clarify them to work alone.

Reviewer 3: lol thanks!

> - C: Once both of the quizzes have gotten all answers at least once each, use `getMoreQuizzes` to load in more quizzes that the user can take.

Reviewer: This one feels much easier than the other two

Author: Clarifying: intent is to get them to have to add the answers for each question to the state, and then be able to read from a current quiz and the previous quiz for knowing whether to call `getMoreQuizzes`.
